### PR TITLE
Kraken: new api ptobject external_code

### PR DESCRIPTION
**Pull request by l-vincent-l**
Will be use by jormungandr to remove the database.

### CODE REVIEW

File: source/kraken/worker.cpp [hunk 1]
```diff
@@ -390,6 +390,90 @@ pbnavitia::Response Worker::place_uri(const pbnavitia::PlaceUriRequest &request)
     return pb_response;
 }
 
+pbnavitia::Response Worker::place_code(const pbnavitia::PlaceCodeRequest &request) {
+    const auto data = data_manager.get_data();
+    this->init_worker_data(data);
+    pbnavitia::Response pb_response;
+
+    auto ext_codes_map = data->pt_data->ext_codes_map[request.type()];
```
Comment (antoine-de) [inline source/kraken/worker.cpp]: you should take a ref on that (a const one)

File: source/kraken/worker.h [hunk 2]
```diff
@@ -84,7 +84,8 @@ class Worker {
         pbnavitia::Response pt_ref(const pbnavitia::PTRefRequest &request);
         pbnavitia::Response disruptions(const pbnavitia::DisruptionsRequest &request);
         pbnavitia::Response calendars(const pbnavitia::CalendarsRequest &request);
-        pbnavitia::Response pt_object(const pbnavitia::PtobjectRequest &request);
+        pbnavitia::Response pt_object(const pbnavitia::PtobjectRequest &request);        
+		pbnavitia::Response place_code(const pbnavitia::PlaceCodeRequest &request);
```
Comment (antoine-de) [inline source/kraken/worker.h]: indent problem

File: source/kraken/worker.cpp [hunk 3]
```diff
@@ -390,6 +390,90 @@ pbnavitia::Response Worker::place_uri(const pbnavitia::PlaceUriRequest &request)
...
+    if (codes_map == ext_codes_map.end()) {
+        fill_pb_error(pbnavitia::Error::unknown_object, "Unknow object", pb_response.mutable_error());
+        return pb_response;
+    }
+    auto uri_it = codes_map->second.find(request.code());
+    if (uri_it == codes_map->second.end()) {
+        fill_pb_error(pbnavitia::Error::unknown_object, "Unknow object", pb_response.mutable_error());
+        return pb_response;
+    }
+    if (request.type() == pbnavitia::PlaceCodeRequest::StopArea) {
```
Comment (antoine-de) [inline source/kraken/worker.cpp]: can't you factorize that ?

File: source/kraken/worker.cpp [hunk 4]
```diff
@@ -512,6 +596,7 @@ pbnavitia::Response Worker::dispatch(const pbnavitia::Request& request) {
         case pbnavitia::METADATAS : return metadatas();
         case pbnavitia::disruptions : return disruptions(request.disruptions());
         case pbnavitia::calendars : return calendars(request.calendars());
+        case pbnavitia::place_code : return place_code(request.place_code()); break;
```
Comment (TeXitoi) [inline source/kraken/worker.cpp]: No useless break please, it's a warning on clang (dead code)

File: source/type/CMakeLists.txt [hunk 5]
```diff
@@ -45,6 +45,7 @@ SET(DATA_SRC
 )
 
 add_library(data ${DATA_SRC})
+target_link_libraries(data types)
```
Comment (TeXitoi) [inline source/type/CMakeLists.txt]: already done in another PR

File: source/type/pb_converter.cpp [hunk 6]
```diff
@@ -722,6 +722,90 @@ void fill_pb_placemark(const navitia::georef::POI* poi,
...
+    if(calendar == nullptr)
+        return;
+    int depth = (max_depth <= 3) ? max_depth : 3;
+    fill_pb_object(calendar, data, place->mutable_calendar(), depth,
+                   now, action_period);
+    place->set_name(calendar->name);
+    place->set_uri(calendar->uri);
+    place->set_embedded_type(pbnavitia::CALENDAR);
+}
+
```
Comment (TeXitoi) [inline source/type/pb_converter.cpp]: All that can be done in one templated method, no?

File: source/type/pt_data.cpp [hunk 7]
```diff
@@ -171,6 +171,30 @@ void PT_Data::build_admins_stop_areas(){
...
+            ext_codes_map[pbnavitia::PlaceCodeRequest::Route][type_code.first][type_code.second] = route->uri;
+    for (auto vehicle_journey: vehicle_journeys)
+        for (auto type_code: vehicle_journey->codes)
+            ext_codes_map[pbnavitia::PlaceCodeRequest::VehicleJourney][type_code.first][type_code.second] = vehicle_journey->uri;
+    for (auto stop_point: stop_points)
+        for (auto type_code: stop_point->codes)
+            ext_codes_map[pbnavitia::PlaceCodeRequest::StopPoint][type_code.first][type_code.second] = stop_point->uri;
+    for (auto calendar: calendars)
+        for (auto type_code: calendar->codes)
+            ext_codes_map[pbnavitia::PlaceCodeRequest::Calendar][type_code.first][type_code.second] = calendar->uri;
```
Comment (TeXitoi) [inline source/type/pt_data.cpp]: Can it be factorised?

File: source/type/pt_data.h [hunk 8]
```diff
@@ -32,22 +32,27 @@ www.navitia.io
...
-// En attendant on utilise std::map car on sait le sérialiser...
-typedef std::map<std::string, idx_t> ExtCodeMap;
+namespace navitia { 
+template <>
+struct enum_size_trait<pbnavitia::PlaceCodeRequest::Type> {
+    static constexpr typename get_enum_type<pbnavitia::PlaceCodeRequest::Type>::type size() {
+        return 8;
+    }
+};
+    namespace type {
```
Comment (TeXitoi) [inline source/type/pt_data.h]: indentation problem

File: source/type/pt_data.h [hunk 9]
```diff
@@ -32,22 +32,27 @@ www.navitia.io
...
 
 struct PT_Data : boost::noncopyable{
 #define COLLECTION_AND_MAP(type_name, collection_name) std::vector<type_name*> collection_name; std::map<std::string, type_name *> collection_name##_map;
     ITERATE_NAVITIA_PT_TYPES(COLLECTION_AND_MAP)
-
+    
+    flat_enum_map<pbnavitia::PlaceCodeRequest::Type, std::unordered_map<std::string, std::map<std::string, std::string>>> ext_codes_map;
```
Comment (TeXitoi) [inline source/type/pt_data.h]: Can the type be explained with typedef? (What are those strings?) And it'll do a smaller line ;)

File: source/type/pt_data.h [hunk 10]
```diff
@@ -32,22 +32,27 @@ www.navitia.io
...
-// Maps pour external code
-// À refléchir si un hash_map ne serait pas mieux ; pas forcément en lecture car hasher une chaîne c'est plus long que comparer
-// En attendant on utilise std::map car on sait le sérialiser...
-typedef std::map<std::string, idx_t> ExtCodeMap;
+namespace navitia { 
+template <>
+struct enum_size_trait<pbnavitia::PlaceCodeRequest::Type> {
+    static constexpr typename get_enum_type<pbnavitia::PlaceCodeRequest::Type>::type size() {
+        return 8;
+    }
```
Comment (TeXitoi) [inline source/type/pt_data.h:47]: Check if the pbnavitia::PlaceCodeRequest::Type are really [0, 8[

File: source/type/request.proto [hunk 11]
```diff
@@ -112,6 +128,7 @@ message Request {
     optional DisruptionsRequest disruptions         = 8;
     optional CalendarsRequest calendars             = 9;
     optional PtobjectRequest pt_objects             = 10;
+	    optional PlaceCodeRequest place_code        = 11;
```
Comment (TeXitoi) [inline source/type/request.proto]: indentation problem

File: source/type/pb_converter.cpp (hunk 6 above)
Comment (l-vincent-l) [inline source/type/pb_converter.cpp]: maybe
